### PR TITLE
Add more contribution docs; closes #912

### LINKED ISSUE #912: Improve documentation that would help potential contributors
I think it would be worthwhile to write down some things that would rule developers. Including the @stylelint/core team and also potential outside contributors. Ideas:

- List commonly overlooked edge cases (e.g. variables, strings & comments, nesting)
- Give examples of different prototypical rules
- Document something about the utils
- Documentation template

**If anybody reading this is a *potential* contributor, what would help you become an *actual* contributor?** (Whether that's docs like I'm talking about or something else?)

### PR DESCRIPTION
**Pull request by davidtheclark**
What do you think of these additions?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,6 +33,20 @@ With stylelint, it's easy to start linting your CSS:
 4. Decide whether to use the [CLI](/docs/user-guide/cli.md), [Node API](/docs/user-guide/node-api.md), or [PostCSS plugin](/docs/user-guide/postcss-plugin.md). Be sure to [specify the syntax](/docs/user-guide/css-processors.md#parsing-scss) if you're using SCSS.
 5. Lint!
 
+## Help out
+
+There is always a lot of work to do, and already well over 100 rules to maintain. So please help out in any way that you can!
+
+You can help stylelint grow in a number of ways:
```
Comment (jeddy3) [inline README.md]: Can we drop "stylelint grow" and just go with: 

"You can help in a number of ways:"

As growth feels like a by-product of making a good tool, rather than a goal in itself.

File: README.md [hunk 2]
```diff
@@ -33,6 +33,20 @@ With stylelint, it's easy to start linting your CSS:
...
+
+- Create, enhance, and debug rules (see our guide to ["Working on rules"](docs/developer-guide/rules.md)).
+- Improve documentation.
+- Chime in on any open issue or pull request.
+- Open new issues about your ideas for making stylelint better.
```
Comment (jeddy3) [inline README.md]: Can we expand on this and encourage experimentation within forks/PRs i.e. rather than just "I have an idea", we encourage "I have an idea and this is how it might work"

Perhaps:

"Open new issues about your ideas for making stylelint better, create a fork and see how it works".

File: README.md [hunk 3]
```diff
@@ -33,6 +33,20 @@ With stylelint, it's easy to start linting your CSS:
...
+- Add new tests to *absolutely anything*.
```
Comment (jeddy3) [inline README.md]: :+1:  :)

File: README.md [hunk 4]
```diff
@@ -33,6 +33,20 @@ With stylelint, it's easy to start linting your CSS:
...
+- Create or contribute to ecosystem tools like the plugins for Atom, Sublime Text, gulp, and Grunt.
```
Comment (jeddy3) [inline README.md]: Shall we drop runners like "gulp" and "grunt" from this list (instead focusing on the editor plugins) until we know the outcome of https://github.com/stylelint/stylelint/issues/917?

File: docs/developer-guide/rules.md [hunk 5]
```diff
@@ -1,57 +1,81 @@
-# Writing rules
+# Working on rules
+
+**Please help us create, enhance, and debug stylelint rules!**
+
+There are well over a hundred rules already, so stylelint *needs* community contributions to continue its growth.
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: Replace "continue its growth", with something like "stay in shape" (for the lack of a better expression)?

File: docs/developer-guide/rules.md [hunk 6]
```diff
@@ -1,57 +1,81 @@
...
+
+If you like stylelint and open source software (since you're reading this, you almost certainly do), please consider taking some time to pitch in. Not only will you help stylelint thrive, you will also learn a thing or two — about CSS, PostCSS, Node, ES2015, unit testing, open source software, and more.
+
+**We want to do everything we can to encourage contributions!** So if you want to participate but don't end up doing it for one reason or another, please file an issue and give us feedback about what we could do to better encourage you.
+
+Also: we hope that your participation in the project isn't a one-off. *We'd love to add more members to the organization and see more regulars pop up in issues and pull requests!*
+
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: :+1: The previous three paragraphs are excellent :)

File: docs/developer-guide/rules.md [hunk 7]
```diff
@@ -1,57 +1,81 @@
...
+## Creating new rules
+
+### Let everyone know
+
+First, open [an issue](https://github.com/stylelint/stylelint/issues/new) and let everyone else know that you intend to create a new rule.
+
+Usually we have some discussion about the rule's purpose, name, and options before it's ready for serious development.
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: Drop "serious"?

File: docs/developer-guide/rules.md [hunk 8]
```diff
@@ -1,57 +1,81 @@
...
+
+Once you have something to show, you'll create a [pull request](https://github.com/stylelint/stylelint/compare) to continue the conversation.
+
+### Come up with a name
 
 Have a look at the [rules guide](/docs/user-guide/rules.md) to familiarize yourself the rule naming conventions.
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: Shall we expand on this?

"Have a look at the [rules guide](/docs/user-guide/rules.md) to familiarize yourself the rule naming conventions. In particular, the section [about rule names](https://github.com/stylelint/stylelint/blob/contribution-docs/docs/user-guide/rules.md#about-rule-names). We take care to ensure that all the rules are named consistently, so that ..." (you said something about this a couple of weeks back, but I can't find it. Something about a good user experience)...

File: docs/developer-guide/rules.md [hunk 9]
```diff
@@ -1,57 +1,81 @@
...
-6. Example patterns that are considered warnings (for each option value).
-7. Example patterns that are *not* considered warnings (for each option value).
-8. Optional options (if applicable).
+*When writing the rule, always look to other similar rules for conventions and patterns to start from and mimic.*
 
-## Single line descriptions
+You will use the simple [PostCSS API](https://github.com/postcss/postcss/blob/master/docs/api.md) to navigate and analyze the CSS syntax tree.
 
-Take the form of:
+Depending on the rule, you may also want ot use [postcss-value-parser](https://github.com/TrySound/postcss-value-parser) and/or [postcss-selector-parser](https://github.com/postcss/postcss-selector-parser), which are easier and more accurate than most people's guesses at regular expressions.
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: typo: "ot"  = "to"

File: docs/developer-guide/rules.md [hunk 10]
```diff
@@ -1,57 +1,81 @@
...
+It is easy to write stylelint tests, so *write as many as you can stand to.*
+
+And please *consider edge-cases.* These are where the bugs and shortcomings of rules always arise.
+
+#### Commonly overlooked edge-cases
+
+- How does your rule handle variables (`$sass`, `@less`, or `var(--custom-property)`)?
+- How does your rule handle CSS strings (e.g. `content: "anything goes";`)?
+- How does your rule handle CSS comments?
+- How does your rule handle nesting?
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: Worth adding these two?:

- How does your rule handle whitespace? (e.g. normalising strings before comparison)
- How does your rule handle data URI schemes?

File: docs/developer-guide/rules.md [hunk 11]
```diff
@@ -1,57 +1,81 @@
...
+You will use the simple [PostCSS API](https://github.com/postcss/postcss/blob/master/docs/api.md) to navigate and analyze the CSS syntax tree.
 
-Take the form of:
+Depending on the rule, you may also want ot use [postcss-value-parser](https://github.com/TrySound/postcss-value-parser) and/or [postcss-selector-parser](https://github.com/postcss/postcss-selector-parser), which are easier and more accurate than most people's guesses at regular expressions.
 
-- "Disallow ..." (for `no` rules).
-- "Limit ..." (for `max` rules).
-- "Require ..." (for `after`, `before` and `inside` rules).
-- "Specify ..." (for everything else).
+stylelint has a number of [utility functions](https://github.com/stylelint/stylelint/tree/master/src/utils) that are used in existing rules and might prove useful to you, as well. Please look through those so that you know what's available. (And if you have a new function that you think might prove generally helpful, let's add it to the list!)
```
Comment (jeddy3) [inline docs/developer-guide/rules.md]: Do we want to single out the `optionsValidator` here, as it's applicable to every rule?

**Comment by jeddy3** [comment]
This looks fantastic to me! 

There's a single typo (hence the "needs revision" flag), and I've made a few suggestions.

**Comment by davidtheclark** [comment]
Thanks, @jeddy3: all good ideas. Please have another look when you feel like it.

**Comment by jeddy3** [comment]
Fab!
